internal/mesh/internal/controllers/proxyconfiguration/controller.go

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package proxyconfiguration
+
+import (
+	"context"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/mesh/internal/controllers/proxyconfiguration/mapper"
+	"github.com/hashicorp/consul/internal/mesh/internal/types"
+	"github.com/hashicorp/consul/internal/resource"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v2beta1"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v2beta1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+const ControllerName = "consul.io/proxy-configuration-controller"
+
+func Controller(proxyConfigMapper *mapper.Mapper) controller.Controller {
+	if proxyConfigMapper == nil {
+		panic("proxy config mapper is required")
+	}
+
+	return controller.ForType(pbmesh.ComputedProxyConfigurationType).
+		WithWatch(pbmesh.ProxyConfigurationType, proxyConfigMapper.MapProxyConfiguration).
```

Comment:
I think you might have the `bound reference` problem here like the routes controller.

1. `ProxyConfiguration` X has 2 selectors (A, B); another Y has the same selectors (A, B)
2. We write 2 `ComputedProxyConfigurations`, one for A and one for B using both (X + Y)
3. User edits X to remove "B" leaving it with just (A, B)
4. We map X and look at the current list which just emits events for [A]
5. We do a no-op reconcile of A.

System final state is:
- A = X+Y **(OK)**
- B = X+Y **(bug)**

Reply:
Yes, great catch! I thought this problem might be hiding somewhere that I missed.

I fixed it in the mapper for now as that's the only place where it makes sense to do for now. I think we will need to address this problem more broadly for cases when we have computed resource and when we don't. Interestingly, gateway api uses the fact that statuses are more or less free-form and stores that info there for parent refs on an HTTP route for example (https://github.com/kubernetes-sigs/gateway-api/blob/d38e6c81ba3d3f137012ceeec8262ab40ad0cd81/apis/v1beta1/shared_types.go#L422 and https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1beta1/httproute_types.go#L1170). This is not the case in our framework today, and I wonder if we should consider statuses to be less strict so that controllers can store info that only they can edit. Status feels better than metadata to me because a user can easily overwrite that.